File: src/eventlog/ieventlog.cc
//=========================================================================
//  IEVENTLOG.CC - part of
//                  OMNeT++/OMNEST
//           Discrete System Simulation in C++
//
//=========================================================================

/*--------------------------------------------------------------*
  Copyright (C) 2006-2017 OpenSim Ltd.

  This file is distributed WITHOUT ANY WARRANTY. See the file
  `license' for details on this and other legal matters.
*--------------------------------------------------------------*/

#include <cstdio>
#include <cstdlib>
#include "ieventlog.h"
#include "index.h"
#include "snapshot.h"

namespace omnetpp {
namespace eventlog {

IEventLog::IEventLog()
{
    clearInternalState();
}

void IEventLog::clearInternalState()
{
    lastNeighbourEventNumber = -1;
    lastNeighbourEvent = nullptr;
}

void IEventLog::synchronize(FileReader::FileChange change)
{
    if (change != FileReader::UNCHANGED) {
        switch (change) {
            case FileReader::OVERWRITTEN:
            case FileReader::APPENDED:
                clearInternalState();
                break;
            default:
                throw opp_runtime_error("Unknown file change");
        }
    }
}

void IEventLog::print(FILE *file, eventnumber_t fromEventNumber, eventnumber_t toEventNumber, bool outputEventLogMessages)
{
    SimulationBeginEntry *simulationBeginEntry = getSimulationBeginEntry();
    if (simulationBeginEntry) {
        simulationBeginEntry->print(file);
        fprintf(file, "\n");
    }
    IEvent *event = fromEventNumber == -1 ? getFirstEvent() : getFirstEventNotBeforeEventNumber(fromEventNumber);
    while (event != nullptr && (toEventNumber == -1 || event->getEventNumber() <= toEventNumber))
    {
        event->print(file, outputEventLogMessages);
        if (event) {
            Index *index = getIndex(event->getEventNumber());
            if (index)
                index->print(file);
            Snapshot *snapshot = getSnapshot(event->getEventNumber());
            if (snapshot)
                snapshot->print(file);
        }
        event = event->getNextEvent();
    }
    SimulationEndEntry *simulationEndEntry = getSimulationEndEntry();
    if (simulationEndEntry)
        simulationEndEntry->print(file);
}

IEvent *IEventLog::getNeighbourEvent(IEvent *event, eventnumber_t distance)
{
    Assert(event);
    eventnumber_t neighbourEventNumber = event->getEventNumber() + distance;

    if (lastNeighbourEvent && lastNeighbourEventNumber != -1 && abs64(neighbourEventNumber - lastNeighbourEventNumber) < abs64(distance))
        return getNeighbourEvent(lastNeighbourEvent, neighbourEventNumber - lastNeighbourEventNumber);

    while (event != nullptr && distance != 0) {
        if (distance > 0) {
            distance--;
            event = event->getNextEvent();
        }
        else {
            distance++;
            event = event->getPreviousEvent();
        }
    }

    lastNeighbourEventNumber = neighbourEventNumber;
    lastNeighbourEvent = (IEvent *)event;

    return lastNeighbourEvent;
}

double IEventLog::getApproximatePercentageForEventNumber(eventnumber_t eventNumber)
{
    IEvent *firstEvent = getFirstEvent();
    IEvent *lastEvent = getLastEvent();
    IEvent *event = getEventForEventNumber(eventNumber);

    if (firstEvent == nullptr || firstEvent == lastEvent)
        return 0.0;
    else if (event == nullptr)
        return 0.5;
    else {
        file_offset_t beginOffset = firstEvent->getBeginOffset();
        file_offset_t endOffset = lastEvent->getBeginOffset();

        double percentage = (double)(event->getBeginOffset() - beginOffset) / (endOffset - beginOffset);
        Assert(0.0 <= percentage && percentage <= 1.0);
        return percentage;
    }
}

}  // namespace eventlog
}  // namespace omnetpp

File: ui/org.omnetpp.common/src/org/omnetpp/eventlog/IEventLog.java
package org.omnetpp.eventlog;

import java.io.OutputStream;
import java.util.TreeSet;

import org.eclipse.core.runtime.IProgressMonitor;
import org.omnetpp.common.engine.BigDecimal;
import org.omnetpp.eventlog.engine.FileReader;
import org.omnetpp.eventlog.entry.SimulationBeginEntry;
import org.omnetpp.eventlog.entry.SimulationEndEntry;

public interface IEventLog
{
    /**
     * Returns the file reader used to read in events.
     */
    public FileReader getFileReader();

    /**
     * Returns the eventlog entry cache for the whole eventlog.
     */
    public EventLogEntryCache getEventLogEntryCache();

    /**
     * Returns the number of events parsed so far.
     */
    public long getNumParsedEvents();

    /**
     * Returns the message names parsed so far.
     */
    public TreeSet<String> getMessageNames();

    /**
     * Returns the message class names parsed so far.
     */
    public TreeSet<String> getMessageClassNames();

    /**
     * Returns the eventlog entry describing the beginning of the simulation.
     */
    public SimulationBeginEntry getSimulationBeginEntry();

    /**
     * Returns the eventlog entry describing the end of the simulation.
     */
    public SimulationEndEntry getSimulationEndEntry();

    /**
     * Returns true if the eventlog does not contain any events.
     */
    public boolean isEmpty();

    /**
     * Returns the first event number or -1 if the log is empty.
     */
    public long getFirstEventNumber();

    /**
     * Returns the first simulation time or 0 if the log is empty.
     */
    public BigDecimal getFirstSimulationTime();

    /**
     * Returns the first event or null if the log is empty.
     */
    public IEvent getFirstEvent();

    /**
     * Returns the last simulation time or 0 if the log is empty.
     */
    public BigDecimal getLastSimulationTime();

    /**
     * Returns the last event number or -1 if the log is empty.
     */
    public long getLastEventNumber();

    /**
     * Returns the last event or null if the log is empty.
     */
    public IEvent getLastEvent();

    /**
     * Returns the requested event or null if there is no such event included in the log.
     * The given event number may not be included in the log.
     */
    public IEvent getEventForEventNumber(long eventNumber);
    public IEvent getEventForEventNumber(long eventNumber, MatchKind matchKind);
    public IEvent getEventForEventNumber(long eventNumber, MatchKind matchKind, boolean useCacheOnly);

    /**
     * Returns the requested event or null if there is no such event included in the log.
     * The given simulation time may not be included in the log.
     */
    public IEvent getEventForSimulationTime(BigDecimal simulationTime);
    public IEvent getEventForSimulationTime(BigDecimal simulationTime, MatchKind matchKind);
    public IEvent getEventForSimulationTime(BigDecimal simulationTime, MatchKind matchKind, boolean useCacheOnly);

    /**
     * Finds the closest eventlog entry containing the given text.
     */
    public EventLogEntry findEventLogEntry(EventLogEntry start, String search, boolean forward, boolean caseSensitive);

    /**
     * Returns the first index or null.
     */
    public Index getFirstIndex();

    /**
     * Returns the last index or null.
     */
    public Index getLastIndex();

    /**
     * Returns the index with the provided event or null if none found.
     * Returns the index that doesn't have the provided event number when searching for the next or the previous.
     */
    public Index getIndex(long eventNumber);
    public Index getIndex(long eventNumber, MatchKind matchKind);

    /**
     * Returns the first snapshot or null.
     */
    public Snapshot getFirstSnapshot();

    /**
     * Returns the last snapshot or null.
     */
    public Snapshot getLastSnapshot();

    /**
     * Returns the snapshot with the provided event number or null if none found.
     * Returns the snapshot that doesn't have the provided event number when searching for the next or the previous.
     */
    public Snapshot getSnapshot(long eventNumber);
    public Snapshot getSnapshot(long eventNumber, MatchKind matchKind);

    /**
     * Returns the approximate number of events present in the log.
     * This value may be less, equal or greater than the real number of events if there are many.
     */
    public long getApproximateNumberOfEvents();

    /**
     * Returns an event approximately at the given percentage in terms of eventlog size.
     */
    public IEvent getApproximateEventAt(double percentage);

    public double getApproximatePercentageForEventNumber(long eventNumber);

    /**
     * Prints the contents into specified file.
     * The given event numbers may not be present in the log.
     */
    public void print();
    public void print(OutputStream stream);
    public void print(OutputStream stream, long fromEventNumber);
    public void print(OutputStream stream, long fromEventNumber, long toEventNumber);
    public void print(OutputStream stream, long fromEventNumber, long toEventNumber, boolean outputEventLogMessages);

    /**
     * Returns the event at the given distance. 0 means the parameter event will be returned.
     */
    public IEvent getNeighbourEvent(IEvent event, long distance);

    /**
     * Returns true if the event with the given event number is included in the log.
     */
    public boolean isIncludedInLog(long eventNumber);

    public IEvent getFirstEventNotBeforeEventNumber(long eventNumber);

    public IEvent getLastEventNotAfterEventNumber(long eventNumber);

    public IEvent getFirstEventNotBeforeSimulationTime(BigDecimal simulationTime);

    public IEvent getLastEventNotAfterSimulationTime(BigDecimal simulationTime);

    public void setProgressMonitor(IProgressMonitor progressMonitor);

    /**
     * Set the minimum interval between progress callbacks for long running eventlog operations.
     */
    public void setProgressCallInterval(double seconds);

    /**
     * Progress notification. May be used to cancel long running operations.
     */
    public void progress();

    /**
     * Synchronizes state when the underlying log file has been changed.
     * Either new events were appended or the whole file has been overwritten.
     */
    public void synchronize(int change);
}
